### DISCUSSION TITLE
Create a filter that uses an external command/script to tell whether to flag a log line

### DISCUSSION THREAD

**Discussion by Occhioverde** [fail2ban/fail2ban#3519] [Q&A] [answered]
Good morning.

Sorry if I'm asking a stupid question, but I couldn't find an answer for it just by looking at the docs: is there a way to create a filter that, instead of passing a log line through a failregex, calls an external executable (or script) that analyzes it in a more complex way than a simple regex and that returns to fail2ban whether it has to issue a ban or not?

I was thinking of writing a small program that checks for suspicious lines, looks for the agents/IPs issuing them in a list of known-malicious ones and then decides if it has to call for a ban, leaving the parsing and banning phases to fail2ban.

I know that I may just create a script that fetches daily the said list of addresses and blocks them directly in iptables, but I thought that, having just a small server that only receives malicious request from a bunch of IPs, downloading and giving the firewall huge lists of addresses that it will likely never see to skim through was a bit overkill (and bad for performance).

I can also make my own tool that parses the logs and maintains its own blocklist in iptables, but I was hoping to leverage the battle-tested and by far more optimized functions of fail2ban.

Thank you in advance for the help.

**Comment by sebres** [comment, marked as answer]
The is no such "filter" directly that would achieve that exactly in this way, but... there are several other ways possible:

1. One could use a simple filter (with regex) triggered by some service with few lines and then use `ignorecommand` to do some more complex analyze (e. g. like you written validate them against some list of known-malicious agents or similar stuff). See https://github.com/fail2ban/fail2ban/issues/2013#issuecomment-426940189 for an example.
With `ignorecache` it'd be also enough performant (called once until result is cached).
Also consider that fail2ban is able to ban everything (not the IPs only), for example the agent or whatever, with appropriate filter and action, like [action.d/nginx-block-map.conf](https://github.com/fail2ban/fail2ban/blob/master/config/action.d/nginx-block-map.conf).
2. Another possibility would be to do all that directly in foreign process and notify fail2ban via `fail2ban-client` or unix socket with fail2ban pickle protocol:
   - either about every failure attempt, see https://github.com/fail2ban/fail2ban/issues/2559#issuecomment-548851675
   - or about a ban (with `fail2ban-client set "$jail" banip ...` or pickle similar example above).
3. Alternatively one'd need to write a new backend as discussed in #881.
